Add Media Foundation decoder

liushuyu opened the pull request:
This PR supersedes #16.

Merged with FFMPEG decoder implementation.

For more details, see #16.
_____
Also, some parts of the code may be reused in Android platform decoding.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/b3n30/citra-1/17)
<!-- Reviewable:end -->

B3n30 commented on src/audio_core/hle/adts.h:
```diff
@@ -0,0 +1,31 @@
+#pragma once
```
Missing license header

B3n30 commented on src/audio_core/hle/adts.h:
```diff
@@ -0,0 +1,31 @@
+
+struct ADTSData {
+    bool MPEG2;
+    uint8_t profile;
```
u8, same for the other types

B3n30 commented on src/audio_core/hle/adts.h:
```diff
@@ -0,0 +1,31 @@
+    uint32_t samplerate;
+};
+
+typedef struct ADTSData ADTSData;
```
I think this isn't necessary in C++

B3n30 replied:
why is this resolved

liushuyu replied:
Sorry, I marked it as resolved since my local commits resolved the issue and I forgot to push them.

B3n30 commented on src/audio_core/hle/adts.h:
```diff
@@ -0,0 +1,31 @@
+#ifdef __cplusplus
+extern "C" {
+#endif // __cplusplus
+uint32_t parse_adts(char* buffer, struct ADTSData* out);
```
I think this can be just a (explicit) constructor of ADTSData, so we can avoid all the extern C stuff.

B3n30 replied:
this isn't resolved

B3n30 commented on src/audio_core/hle/adts_reader.c:
```diff
@@ -0,0 +1,49 @@
+
```
Citra header missing

B3n30 commented on src/audio_core/hle/adts_reader.c:
```diff
@@ -0,0 +1,49 @@
+
+#include "adts.h"
+
+const uint32_t freq_table[16] = {96000, 88200, 64000, 48000, 44100, 32000, 24000, 22050,
```
constexpr std::array<u32, 16> freq_table =

B3n30 commented on src/audio_core/hle/adts_reader.c:
```diff
@@ -0,0 +1,49 @@
+}
+
+// last two bytes of MF AAC decoder user data
+uint16_t mf_get_aac_tag(struct ADTSData input) {
```
u16

B3n30 commented on src/audio_core/hle/adts_reader.c:
```diff
@@ -0,0 +1,49 @@
+const short channel_table[8] = {0, 1, 2, 3, 4, 5, 6, 8};
+
+uint32_t parse_adts(char* buffer, struct ADTSData* out) {
+    uint32_t tmp = 0;
```
u32

B3n30 commented on src/audio_core/hle/adts_reader.c:
```diff
@@ -0,0 +1,49 @@
+
+const uint32_t freq_table[16] = {96000, 88200, 64000, 48000, 44100, 32000, 24000, 22050,
+                                 16000, 12000, 11025, 8000,  7350,  0,     0,     0};
+const short channel_table[8] = {0, 1, 2, 3, 4, 5, 6, 8};
```
`constexpr std::array<u8, 8> channel_table = `

B3n30 commented on src/audio_core/hle/wmf_decoder.cpp:
```diff
@@ -0,0 +1,254 @@
+
+    Memory::MemorySystem& memory;
+
+    IMFTransform* transform = NULL;
```
nullptr, This should probably be a unique_ptr with proper DTOR similar to what was done in the ffmpeg_decoder, to assure it always gets freed

liushuyu replied:
Not possible. WINAPIs do not conform to standard C++ RAII. Interfaces do not come with a destructor that frees resources. MS documents suggest `->Release()` is the only way to free the resources.

B3n30 replied:
Thats exactly what i meant inside the unique_ptr dtor the `->Release()` should get called

B3n30 commented on src/audio_core/hle/wmf_decoder.cpp:
```diff
@@ -0,0 +1,254 @@
+    selected = false;
+}
+
+int WMFDecoder::Impl::DecodingLoop(ADTSData adts_header,
```
`std::optional<std::array<std::vector<u8>, 2>> WMFDecoder::Impl::DecodingLoop(const ADTSData& adts_header)`

liushuyu replied:
So how do I check if the `DecodingLoop` succeeded w/o status code? It's not like MFT will always return something if it returns `S_OK`

B3n30 replied:
with the optional. if optional has a value it succeed if not it failed. You never check what the failed error code is, nevertheless

liushuyu replied:
>with the optional. if optional has a value it succeed if not it failed. You never check what the failed error code is, nevertheless

Also, you should always check the error code __despite__ if there is any output you received. This is a strange behavior I have observed. When MFT returned a `MORE_DATA_AVAILABLE`, sometimes you will get a `NULL` next time you try to receive the sample and MFT will tell you the operation succeeded with 0 length of data.

B3n30 commented on src/audio_core/hle/wmf_decoder.cpp:
```diff
@@ -0,0 +1,254 @@
+int WMFDecoder::Impl::DecodingLoop(ADTSData adts_header,
+                                   std::array<std::vector<u8>, 2>& out_streams) {
+    int output_status = 0;
+    char* output_buffer = NULL;
```
nullptr

B3n30 commented on src/audio_core/hle/wmf_decoder.cpp:
```diff
@@ -0,0 +1,254 @@
+    int output_status = 0;
+    char* output_buffer = NULL;
+    DWORD output_len = 0;
+    IMFSample* output = NULL;
```
nullptr

B3n30 commented on src/audio_core/hle/wmf_decoder.cpp:
```diff
@@ -0,0 +1,254 @@
+                    out_streams[1].data(), out_streams[1].size());
+    }
+
+    response.num_channels = adts_header.channels;
```
Probably move this and the adts_header parsing before the init check

liushuyu replied:
I'll address this

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+    }
+
+    ADTSData tmp;
+    UINT8 aac_tmp[] = {0x01, 0x00, 0xfe, 00, 00, 00, 00, 00, 00, 00, 00, 00, 0x00, 0x00};
```
why isn't aac_tmp just a member of the ADTS_header and the parsing of that takes place in parse_adts?

aac_tag should probably be a std::array<>

liushuyu replied:
I want to reuse this parser for Android AMedia framework. The decoder in Android also needs you to parse the ADTS for it.

liushuyu replied:
This thing is only needed for MF so I will keep this part here.

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+    return -1;
+}
+
+int detect_mediatype(char* buffer, size_t len, ADTSData* output, char** aac_tag) {
```
`bool DetectMediatype` also all the pointer could just be references, or maybe avoid this function completely bu juts using a ADTS constructor

B3n30 commented on src/audio_core/hle/wmf_decoder.cpp:
```diff
@@ -0,0 +1,254 @@
+        if (output_status == 3)
+            continue;
+
+        return output_status; // return on other status
```
LOG_ERROR the output_status in this case?

liushuyu replied:
Addressed

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+    LPSTR err;
+    FormatMessage(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_ALLOCATE_BUFFER |
+                      FORMAT_MESSAGE_IGNORE_INSERTS,
+                  NULL, hr,
```
nullptr

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+                  // hardcode to use en_US because if any user had problems with this
+                  // we can help them w/o translating anything
+                  MAKELANGID(LANG_ENGLISH, SUBLANG_ENGLISH_US), (LPSTR)&err, 0, NULL);
+    if (err != NULL) {
```
nullptr

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+                  NULL, hr,
+                  // hardcode to use en_US because if any user had problems with this
+                  // we can help them w/o translating anything
+                  MAKELANGID(LANG_ENGLISH, SUBLANG_ENGLISH_US), (LPSTR)&err, 0, NULL);
```
nullptr

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+    LOG_CRITICAL(Audio_DSP, "{}: {:08x}", msg, hr);
+}
+
+int mf_coinit() {
```
`bool MFCoinit()`

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+    return 0;
+}
+
+int mf_decoder_init(IMFTransform** transform, GUID audio_format) {
```
`bool MFDecoderInit`

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+    for (unsigned int n = 0; n < num_activate; n++) {
+        hr = activate[n]->ActivateObject(IID_IMFTransform, (void**)transform);
+        if (FAILED(hr))
+            *transform = NULL;
```
nullptr

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+            *transform = NULL;
+        activate[n]->Release();
+    }
+    if (*transform == NULL) {
```
nullptr, dito for below

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+    return 0;
+}
+
+void mf_deinit(IMFTransform** transform) {
```
MFDeinit

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+    CoUninitialize();
+}
+
+IMFSample* create_sample(void* data, DWORD len, DWORD alignment, LONGLONG duration) {
```
CreateSample

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+    return sample;
+}
+
+int select_input_mediatype(IMFTransform* transform, int in_stream_id, ADTSData adts,
```
`bool SelectInputMediaType(IMFTransform* transform, int in_stream_id, const ADTSData& adts,`

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+    return 0;
+}
+
+int select_output_mediatype(IMFTransform* transform, int out_stream_id, GUID audio_format) {
```
`bool SelectOutputMediatype(..., const GUID& audio_format)`

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+    return 0;
+}
+
+int mf_flush(IMFTransform** transform) {
```
the return isn't necessary also `MFFlush`

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+    return 0;
+}
+
+int send_sample(IMFTransform* transform, DWORD in_stream_id, IMFSample* in_sample) {
```
`bool SendSample`

liushuyu replied:
No. This one is a tri-state function.

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+        } // FAILED(hr)
+    } else {
+        hr = transform->ProcessMessage(MFT_MESSAGE_COMMAND_DRAIN, 0);
+        // ffmpeg: Some MFTs (AC3) will send a frame after each drain command (???), so
```
why `ffmpeg:`?

liushuyu replied:
This function is taken from a proposed patch of `ffmpeg`. While this is not the case with AAC, I figured I should leave the comment here.

liushuyu replied:
I think I don't need to do anything for this one. Marking as resolved

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+    return 0;
+}
+
+// return: 0: okay; 1: needs more sample; 2: needs reconfiguring; 3: more data available
```
could be an enum?

liushuyu replied:
Addressed with the usage of
 ```c
enum MFOutputState
```

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+}
+
+// return: 0: okay; 1: needs more sample; 2: needs reconfiguring; 3: more data available
+int receive_sample(IMFTransform* transform, DWORD out_stream_id, IMFSample** out_sample) {
```
ReceiveSample

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+        return 3;
+    }
+
+    // TODO: better handling try again and EOF cases using drain value
```
TODO(liushuyu):

liushuyu replied:
I just removed the comment since AAC decoder is a `SYNCMFT` it does not need to keep track of drain status if you are doing the drain "correctly"

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.cpp:
```diff
@@ -0,0 +1,366 @@
+    return 0;
+}
+
+int copy_sample_to_buffer(IMFSample* sample, void** output, DWORD* len) {
```
why not return the output (maybe as std::vector)? And `CopySampleToBuffer`

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.h:
```diff
@@ -0,0 +1,48 @@
+#ifndef MF_DECODER
+#define MF_DECODER
+
+#define WINVER _WIN32_WINNT_WIN7
```
Add a comment why this is necessary

liushuyu replied:
Comment added

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.h:
```diff
@@ -0,0 +1,48 @@
+
+// utility functions
+template <class T>
+void SafeRelease(T** ppT) {
```
This could be avoided by using unique_ptr and would also "SafeRelease" in case of an exception

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.h:
```diff
@@ -0,0 +1,48 @@
+#pragma once
```
missing citra header

B3n30 commented on src/audio_core/hle/wmf_decoder_utils.h:
```diff
@@ -0,0 +1,48 @@
+#pragma once
+
+#ifndef MF_DECODER
```
this shouldn't be necessary

B3n30 commented on src/tests/audio_core/audio_fixures.h:
```diff
@@ -0,0 +1,5 @@
+const int fixure_buffer_size = 41;
```
constexpr

B3n30 commented on src/tests/audio_core/audio_fixures.h:
```diff
@@ -0,0 +1,5 @@
+const int fixure_buffer_size = 41;
+const unsigned char fixure_buffer[41] = {
```
constexpr std::array<u8, 41>

B3n30 commented on src/tests/audio_core/decoder_tests.cpp:
```diff
@@ -0,0 +1,58 @@
+// Copyright 2017 Citra Emulator Project
```
2019

B3n30 commented on src/audio_core/CMakeLists.txt:
```diff
@@ -28,7 +28,7 @@ add_library(audio_core STATIC
     $<$<BOOL:${ENABLE_CUBEB}>:cubeb_sink.cpp cubeb_sink.h>
     $<$<BOOL:${FFMPEG_FOUND}>:hle/ffmpeg_decoder.cpp hle/ffmpeg_decoder.h hle/ffmpeg_dl.cpp hle/ffmpeg_dl.h>
-    $<$<BOOL:${ENABLE_MF}>:hle/wmf_decoder.cpp hle/wmf_decoder.h hle/wmf_decoder_utils.cpp hle/wmf_decoder_utils.h hle/adts_reader.c>
+    $<$<BOOL:${ENABLE_MF}>:hle/wmf_decoder.cpp hle/wmf_decoder.h hle/wmf_decoder_utils.cpp hle/wmf_decoder_utils.h hle/adts_reader.cpp>
```
Imo the adts_reader stuff isn't for wmf only. My plan is to use it inside the NullDecoder too, so we could give a proper response there. So I think those files can be outside of the BOOL part. I also didn't check it properly yet, but isn't hle/adts.h missing?

liushuyu commented:
I have addressed most reviews. The things left are:

- Use smart pointers instead of raw pointers
- Return `std::vector` or `std::array` instead of pointer de-reference

I really not good at writing modern C++ code, if you could help me with these, I will be more than happy to learn how you do these things correctly.

Thanks.
